mat/matrix.go

```diff
@@ -846,6 +846,21 @@ func normLapack(norm float64, aTrans bool) lapack.MatrixNorm {
 	}
 }
 
+// Rank returns rank of input matrix using singular value decomposition.
```

Comment:
Thank you for reviewing. Forgive me for slow working.

> If epsilon is negative, Rank will just return min(m, n). This will behave as documented but it will not return the rank of A (if different from min(m,n)). Do we want to allow this or panic instead?

I added panic for negative epsilon for safety. Thank you for catching.

> Do we want to replace zero epsilon with some default value? I guess it is difficult to have an intuition for choosing a proper value of epsilon, so it could be a convenient tool which on the other hand adds (small) complexity that would have to be documented.

Replacing default value is a good idea (in my original implementation, epsilon was global variable to have default value). Since epsilon is often used as positive (and small) number, replacing zero-epsilon with default value may not be counterintuitive.
Could you please review additional commit?

Reply:
May be my epsilon is too large?